benchmarks/bench_multi_gpu.py

```diff
@@ -0,0 +1,241 @@
+        rank
+    )
+    ref_optimizer = torch.optim.SGD(ref_model.parameters(), lr=lr * world_size)
+    if compile:
+        ref_model = torch.compile(ref_model)
+    ref_model = FSDP(ref_model, use_orig_params=compile)
+
+    def ref_forw_backward():
+        ref_optimizer.zero_grad()
+        ref_model(input_tensor).sum().backward()
+        ref_optimizer.step()
+
+    if transformer_engine_installed:
+        te_model = FSDP(
+            get_model(K, N, is_fp8=False, is_te=True, base_dtype=base_dtype).to(rank),
+            use_orig_params=compile,
+        )
+        fp8_format = recipe.Format.HYBRID
+        fp8_recipe = recipe.DelayedScaling(
+            fp8_format=fp8_format, amax_history_len=16, amax_compute_algo="max"
+        )
+        # Compiling TE_linear fails but they are already compiling under the hood
+        # if transformer_engine_installed:
+        #     te_forw_backward = torch.compile(te_forw_backward)
+        if rank == 0:
+            print(te_model)
+
+        te_optimizer = torch.optim.SGD(ref_model.parameters(), lr=lr * world_size)
+
+        def te_forw_backward():
+            te_optimizer.zero_grad()
+            with te.fp8_autocast(enabled=True, fp8_recipe=fp8_recipe):
+                y = te_model(input_tensor)
+            y.sum().backward()
+            te_optimizer.step()
+
+    def run_n_iterations(n, fn):
+        for _ in range(n):
+            fn()
+        # make sure training is done on all ranks
+        dist.barrier()
+
+    # warmup
+    run_n_iterations(50, ref_forw_backward)
+    run_n_iterations(50, float8_forw_backward)
+    if transformer_engine_installed:
+        run_n_iterations(50, te_forw_backward)
+
+    N_ITER = 50
+    ref_time = (
+        benchmark_torch_function_in_microseconds(
+            run_n_iterations, N_ITER, ref_forw_backward
+        )
+        * 1e-6
+        / N_ITER
+    )
+    float8_time = (
+        benchmark_torch_function_in_microseconds(
+            run_n_iterations, N_ITER, float8_forw_backward
+        )
+        * 1e-6
+        / N_ITER
+    )
+    if transformer_engine_installed:
+        te_time_sec = (
+            benchmark_torch_function_in_microseconds(
+                run_n_iterations, N_ITER, te_forw_backward
+            )
+            * 1e-6
+            / N_ITER
+        )
+    else:
+        te_time_sec = None
+
+    if rank == 0:
+        print("ref_time", ref_time)
+        print("float8_time", float8_time)
+        print("te_time_sec", te_time_sec)
+        print("float8 speedup", ref_time / float8_time)
+        if transformer_engine_installed:
+            print("te speedup", ref_time / te_time_sec)
+
+    cleanup()
+
+
+def run():
+    compile = True
+    base_dtype = torch.bfloat16
+    print(f"{base_dtype = }")
+    print(f"{compile = }")
+    # generate input data
+    ref_input = torch.randn(B, M, K).cuda().to(base_dtype)
+    # run fsdp model
+    WORLD_SIZE = torch.cuda.device_count()
```

Comment:
maybe make this configurable? usually easier to debug issues and iterate with world_size == 2, and then verify the fix with world_size == torch.cuda.device_count()